### PR TITLE
Accommodate upstream changes to KVCache

### PR DESCRIPTION
**Pull request by uogbuji**
Hi, I created another open-source project based on this one, Toolio (in which I credit upstream, of course). It has some code refactored from here, and I ran into an issue with `ReusableKVCache` based on recent modifications in [mlx-examples](https://github.com/ml-explore/mlx-examples/):

* https://github.com/ml-explore/mlx-examples/commit/69181e00584976f717c27f90c7e009e70cc1b0bf
* https://github.com/ml-explore/mlx-examples/commit/561dcf5643ecced3be9bf053499e0e2ea51e8686

i'm still grokking all this, so the PR might need a tweak, but I didn't want to sit on it when there could be discussion.

Thanks for llm-structured-output! Awesome project!

### CODE REVIEW

File: src/examples/reusable_kv_cache.py [hunk 1]
```diff
@@ -48,12 +49,16 @@ def update_and_fetch(self, keys, values):
...
             n_steps = (self.step + keys.shape[2] - 1) // self.step
-            shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.head_dim)
-            new_k = mx.zeros(shape, keys.dtype)
-            new_v = mx.zeros(shape, values.dtype)
+            k_shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.k_head_dim)
+            v_shape = (keys.shape[0], self.n_kv_heads, n_steps * self.step, self.v_head_dim)
+            new_k = mx.zeros(k_shape, keys.dtype)
+            new_v = mx.zeros(v_shape, values.dtype)
             if self.keys is not None:
+                if (prev + keys.shape[2]) > self.keys.shape[2]:
```
Comment (otriscon) [inline src/examples/reusable_kv_cache.py]: In `mlx-examples`, the corresponding line (#38) in method `update_and_fetch` of `mlx_lm/models/base.py` is instead:

`if prev % self.step != 0:`

Did you find that changing this line was necessary for it to work well?

My concern with this change is that it may resize the KV cache more often than necessary, affecting performance.

**Comment by otriscon** [comment]
Hi @uogbuji. Thank you for submitting this PR, and congratulations on Toolio, it looks really cool.

I've reviewed the change and added a comment.

Can I ask you to please also modify `src/examples/requirements.txt` so that `mlx-lm>=0.16`? Since this is a breaking change for people with a previous version of `mlx-lm` installed, it will at least hint at the cause and provide an easy recovery.

I've successfully tested the change by running an evaluation with `mlx-community/Meta-Llama-3-8B-Instruct-8bit` for backward compatibility.

I've tested with `deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct` as well but it's not working yet. There seems to be a mismatch in the shape of the queries matrix. I'm looking into it.

**Comment by otriscon** [comment]
The issue that we're running into with DeepSeek is that the way the Deepseek model implementation initializes the attention mask in `mlx-lm` is not cache-reuse friendly.

The reason `ReusableKVCache` exists is to accelerate successive similar requests by reusing the common parts of the prompt. This is a common pattern in evaluation in our experience and leads to large improvements in time to first token.

The way it works is that, on the first evaluation, the KV cache operates normally and stores the key and value matrices for each layer. First the whole prompt is evaluated in one go (prefill stage) and then we generate tokens one by one.

When the next evaluation starts, we compare the prompt with the prompt from the prior evaluation. If they have prefix tokens in common, we can reuse part of the matrices in the evaluation of the second prompt.

Let's say the first evaluation has concluded with a total of 1000 tokens including both prompt and generated tokens, and the first 300 tokens of the second prompt are exactly the same as for the first prompt. We can reuse 300 columns of each of the key and value matrices for each layer in the model. Now, we need to evaluate (prefill) the rest of the second prompt (the part that is not common with the first), and here's where we're running into a problem.

In the prefill stage, where we go over multiple tokens at once, a causal mask must be applied so that the attention mechanism doesn't "look ahead" at future tokens. The dimensions of this mask need to match the dimensions of the input.

Looking into `mlx-examples/llms/mlx_lm/models/deepseek_v2.py` we see that there is an implicit assumption that if the input contains more than one token, then there aren't any prior tokens and the mask is created accordingly:

```
class DeepseekV2Model(nn.Module):
    ... omitted for clarity
            
    def __call__(
        self,
        x: mx.array,
        cache: Optional[KVCache] = None,
    ) -> mx.array:
        h = self.embed_tokens(x)
        mask = None
        T = h.shape[1]
        if T > 1:
            mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
            mask = mask.astype(h.dtype)

        ...        
```

This used to be the case for other model implementations in `mlx-lm`, but a while ago `create_additive_causal_mask` was added to cover for this. So there's two solutions:

1. The trivial but not very satisfactory solution is to not use prompt reuse for this model by passing `cache_prompt=False` to the `completion` method of `llm_schema`.

2. The better solution is to submit an upstream PR to `mlx-lm` with the following changes to `deepseek_v2.py`:

```
Line 8:
< from .base import BaseModelArgs, KVCache
> from .base import BaseModelArgs, KVCache, create_additive_causal_mask
 
Line 414:
<            mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
>            mask = create_additive_causal_mask(
>                T, cache[0].offset if cache is not None else 0
>            )
```

In either case, I think both are out of the scope of this PR.

File: src/examples/reusable_kv_cache.py (hunk 1 above)
Comment (uogbuji) [reply src/examples/reusable_kv_cache.py]: Hmm. I actually changed it to that modulus line, and changed it back to the one you flagged. But…I can't recall why I did that. 😂

I'll change it back. Your instincts are almost certainly more reliable than my fiddling.

**Comment by uogbuji** [comment]
> Hi @uogbuji. Thank you for submitting this PR, and congratulations on Toolio, it looks really cool.

Thanks! It has been hugely fun to work with. An initial confessionL I haven't been able to get truly into the heart of the model sampling/steering. I've mostly been building a useful scaffolding around your brilliant black box. I do plan to really dig in one of these weekends. Until then, please forgive any naivety on my part.

> I've reviewed the change and added a comment.
> 
> Can I ask you to please also modify `src/examples/requirements.txt` so that `mlx-lm>=0.16`? Since this is a breaking change for people with a previous version of `mlx-lm` installed, it will at least hint at the cause and provide an easy recovery.

Yes. Good point!

> I've successfully tested the change by running an evaluation with `mlx-community/Meta-Llama-3-8B-Instruct-8bit` for backward compatibility.
> 
> I've tested with `deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct` as well but it's not working yet. There seems to be a mismatch in the shape of the queries matrix. I'm looking into it.
…
> The issue that we're running into with DeepSeek is that the way the Deepseek model implementation initializes the attention mask in `mlx-lm` is not cache-reuse friendly.
> 
> The reason `ReusableKVCache` exists is to accelerate successive similar requests by reusing the common parts of the prompt. This is a common pattern in evaluation in our experience and leads to large improvements in time to first token.

Yeah, definitely get the motivation.

> The way it works is that, on the first evaluation, the KV cache operates normally and stores the key and value matrices for each layer. First the whole prompt is evaluated in one go (prefill stage) and then we generate tokens one by one.
> 
> When the next evaluation starts, we compare the prompt with the prompt from the prior evaluation. If they have prefix tokens in common, we can reuse part of the matrices in the evaluation of the second prompt.
> 
> Let's say the first evaluation has concluded with a total of 1000 tokens including both prompt and generated tokens, and the first 300 tokens of the second prompt are exactly the same as for the first prompt. We can reuse 300 columns of each of the key and value matrices for each layer in the model. Now, we need to evaluate (prefill) the rest of the second prompt (the part that is not common with the first), and here's where we're running into a problem.
> 
> In the prefill stage, where we go over multiple tokens at once, a causal mask must be applied so that the attention mechanism doesn't "look ahead" at future tokens. The dimensions of this mask need to match the dimensions of the input.
> 
> Looking into `mlx-examples/llms/mlx_lm/models/deepseek_v2.py` we see that there is an implicit assumption that if the input contains more than one token, then there aren't any prior tokens and the mask is created accordingly:
> 
> ```
> class DeepseekV2Model(nn.Module):
>     ... omitted for clarity
>             
>     def __call__(
>         self,
>         x: mx.array,
>         cache: Optional[KVCache] = None,
>     ) -> mx.array:
>         h = self.embed_tokens(x)
>         mask = None
>         T = h.shape[1]
>         if T > 1:
>             mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
>             mask = mask.astype(h.dtype)
> 
>         ...        
> ```
> 
> This used to be the case for other model implementations in `mlx-lm`, but a while ago `create_additive_causal_mask` was added to cover for this. So there's two solutions:
> 
> 1. The trivial but not very satisfactory solution is to not use prompt reuse for this model by passing `cache_prompt=False` to the `completion` method of `llm_schema`.

Yeah, I actually did this in Toolio.

> 2. The better solution is to submit an upstream PR to `mlx-lm` with the following changes to `deepseek_v2.py`:
> 
> ```
> Line 8:
> < from .base import BaseModelArgs, KVCache
> > from .base import BaseModelArgs, KVCache, create_additive_causal_mask
>  
> Line 414:
> <            mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
> >            mask = create_additive_causal_mask(
> >                T, cache[0].offset if cache is not None else 0
> >            )
> ```
> 
> In either case, I think both are out of the scope of this PR.

I agree. Any chance you might have a moment to submit that upstream PR, since as I've confessed, you'd be more knowledgeable than I am? I did appreciate you really clear explanation, though!

I'll make the modulus & version pin changes to this PR. Thanks for the quick review. I look forward to continued collab!

**Comment by otriscon** [comment]
Thanks @uogbuji.

> An initial confessionL I haven't been able to get truly into the heart of the model sampling/steering.

If you ever have the time to go over it, hit me up and I can walk you through it and answer any questions.

Thank you for making the changes. Can I be nit-picky and ask you to please remove the comment in this line:

```
    if prev % self.step != 0: # Resize when we hit a step
```

The reason is that I'd like it to match the upstream line which doesn't have the comment. That way it's easier to track what's different and keep it in sync. Sorry for the hassle.

> Any chance you might have a moment to submit that upstream PR

Sure. I see that there's quite a few models that have the same issue so it may be better to fix them all at once and it may need a bit more work, so I may not get to it today.

**Comment by uogbuji** [comment]
> Thanks @uogbuji.
> 
> > An initial confessionL I haven't been able to get truly into the heart of the model sampling/steering.
> 
> If you ever have the time to go over it, hit me up and I can walk you through it and answer any questions.

Let me get a bit more of a messy hands-on with transformers this weekend, then I'll take you up on that!

> Thank you for making the changes. Can I be nit-picky and ask you to please remove the comment in this line:
> 
> ```
>     if prev % self.step != 0: # Resize when we hit a step
> ```
> 
> The reason is that I'd like it to match the upstream line which doesn't have the comment. That way it's easier to track what's different and keep it in sync. Sorry for the hassle.

Aye. Done!

> > Any chance you might have a moment to submit that upstream PR
> 
> Sure. I see that there's quite a few models that have the same issue so it may be better to fix them all at once and it may need a bit more work, so I may not get to it today.

Understood.

**Review by otriscon — APPROVED** [review]
